File: frontend/src/features/crm/types/sales-order.ts
export type OrderStatus =
  | "DRAFT"
  | "PENDING"
  | "APPROVED"
  | "PROCESSING"
  | "COMPLETED"
  | "CANCELLED";

export type OrderType = "STANDARD" | "SERVICE" | "SUBSCRIPTION" | string;

export type OrderItem = {
  id?: number;
  itemName: string | null;
  sku: string | null;
  description: string | null;
  quantity: number | null;
  unitPrice: number | null;
  totalPrice?: number | null;
};

export type SalesOrder = {
  id: number;
  orderNumber: string;
  customerId: number;
  opportunityId: number | null;
  quotationId: number | null;
  status: OrderStatus;
  type: OrderType | null;
  orderDate: string | null;
  deliveryDate: string | null;
  subtotal: number | null;
  taxAmount: number | null;
  discountAmount: number | null;
  totalAmount: number | null;
  currency: string | null;
  notes: string | null;
  items: OrderItem[] | null;
  createdAt: string | null;
  updatedAt: string | null;
};

File: services/crm-service/src/main/java/com/crm/crm_backend/model/entity/SalesOrder.java
package com.crm.crm_backend.model.entity;

import com.crm.crm_backend.model.enums.OrderStatus;
import com.crm.crm_backend.model.enums.OrderType;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "sales_orders")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class SalesOrder {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "order_number", unique = true, nullable = false)
    private String orderNumber;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id")
    private Customer customer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "opportunity_id")
    private Opportunity opportunity;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "quotation_id")
    private Quotation quotation;

    @Enumerated(EnumType.STRING)
    private OrderStatus status;

    @Enumerated(EnumType.STRING)
    private OrderType type;

    @Column(name = "order_date")
    private LocalDate orderDate;

    @Column(name = "delivery_date")
    private LocalDate deliveryDate;

    @Column(precision = 15, scale = 2)
    private BigDecimal subtotal;

    @Column(precision = 15, scale = 2)
    private BigDecimal taxAmount;

    @Column(precision = 15, scale = 2)
    private BigDecimal discountAmount;

    @Column(precision = 15, scale = 2)
    private BigDecimal totalAmount;

    private String currency;

    @Column(length = 2000)
    private String notes;

    @OneToMany(
            mappedBy = "salesOrder",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    @Builder.Default
    private List<OrderItem> orderItems = new ArrayList<>();

    private LocalDateTime createdAt;

    private LocalDateTime updatedAt;

    @PrePersist
    public void prePersist() {

        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();

        if (status == null) {
            status = OrderStatus.DRAFT;
        }

        if (currency == null) {
            currency = "ETB";
        }
    }

    @PreUpdate
    public void preUpdate() {
        updatedAt = LocalDateTime.now();
    }
}
